### DISCUSSION TITLE
Mobile/Web Relays

### DISCUSSION THREAD

**Discussion by SeanCheatham** [GiraffeChain/giraffe#38] [Ideas]
Giraffe currently supports staking and block production from a mobile/web context, but it still relies heavily on a connection to a trusted relay node.

The project didn't start that way though. It actually started entirely in Dart/Flutter. All relay operations (validation and P2P) were natively implemented in Dart. I was able to get it to work, at least on mobile. But there were a few obstacles that made me pivot direction.
1. Firewalls and NAT currently make mobile P2P impractical. As far as I know, there are no 3G/4G/5G cell carriers that allow inbound connections to your device.
1. Mobile and web data storage are constrained
1. CPU and memory are constrained

But I have a few thoughts on how to overcome those challenges:
1. WebRTC is generally intended for media streaming, like video conferencing. But a less-mentioned aspect is its ability to support a P2P mesh network by using a server to facilitate new connections between clients. If Giraffe's P2P layer could be adapted to UDP, then STUN could _possibly_ be leveraged to allow direct mobile-to-mobile connections. Alternatively, it could stick with TCP and use TURN, which acts more like a middleman between two peers.
1. A relay doesn't need to store the entire chain. It could instead allocate a certain amount of space, and sparsely store random sections of the chain. For any missing data, just ask around to other peers. Hopefully someone has what's needed.
1. After experimenting with a couple testnets over the last several weeks, I'm a lot less concerned about this than I was before. While the peer graph is relatively small, it is geographically distributed between the USA, Europe, and Australia. On a 1GB/1vCPU DigitalOcean instance, CPU usage sits at around 2%. A staker on a 0.5GB/1vCPU instance consumes around 15% CPU. Higher transaction throughput will certainly increase those numbers, but the core consensus consumes hardly a thing.

So I think it might be time to reconsider running full relays on mobile phones and in web browsers. What do you think?

**Comment by SeanCheatham** [comment]
An advantage I see to including mobile phones in the mix: CDN / Edge Storage

A network of millions of phones can help distribute data in a relatively efficient manner. A block produced in Australia may need to be delivered to thousands of people in New York City. Rather than transferring the block across the ocean thousands of times, it could be transferred once to a single person, and that single person could rebroadcast it to the rest of the city.

It becomes even more interesting if we leverage phone GPS to help assist with optimizing the P2P graph. That said, it scares the heck out of me to try to incorporate location tracking into this protocol. It would _absolutely need to be opt-in_.

**Comment by mgrand** [comment]
It sounds like you want to build a distributed hash map. I'm not aware of
any that use physical location to organize. The ones I have seen use
network topology or are random. It will be interesting to see how much
awareness of physical location helps or doesn't.

You might want to take a look at the Kademlia Distributed Hash Table (DHT)
Algorithm.

One interesting thing about this organization is that not every node in the
network needs to have every block or other chunk of data in its own
storage. You just need a guarantee that every block is stored on at least
one node. Each block is uniquely identified by its hash. Instead of trying
to send blocks synchronously, they can be fetched on demand from the
nearest place they happen to be.

On Sat, Oct 5, 2024 at 9:29 AM Sean Cheatham ***@***.***>
wrote:

> An advantage I see to including mobile phones in the mix: CDN / Edge
> Storage
>
> A network of millions of phones can help distribute data in a relatively
> efficient manner. A block produced in Australia may need to be delivered to
> thousands of people in New York City. Rather than transferring the block
> across the ocean thousands of times, it could be transferred once to a
> single person, and that single person could rebroadcast it to the rest of
> the city.
>
> It becomes even more interesting if we leverage phone GPS to help assist
> with optimizing the P2P graph. That said, it scares the heck out of me to
> try to incorporate location tracking into this protocol. It would *absolutely
> need to be opt-in*.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/GiraffeChain/giraffe/discussions/38#discussioncomment-10852436>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEUGFUTUZAJEC2FCTU5EK3ZZ7SUXAVCNFSM6AAAAABPIVYE6KVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTAOBVGI2DGNQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***
> com>
>


-- 
Mark Grand
(404)925 8265
Vision, Values, Verve:
* Visualize the end game with
* Values to lead ethically and with purpose; and
* Verve to execute with shameless audacity.

**Reply by SeanCheatham** [reply]
:partying_face: Yes! I'm basically hoping to achieve a DHT where the data is block/transaction information.

I hadn't heard of Kademlia before but it looks super interesting! From my understanding, peers pick a random location within the hash table and store data with hashes "near" that location. Peers keep track of other peers and the sections of the hash table they are likely to store, but they bias toward tracking peers that are similar in the hash table to themselves. When requesting data that is close-by (from a hash perspective), it's very easy because you're so knowledgeable of nearby peers. When requesting data that's far away in the hash table, you find a peer in the rough ballpark and ask them for help because _they_ know the area better than you. It might take multiple hops, but theoretically it should never take very many.

Perhaps one way to tune Kademlia toward physical location is in the peer biasing.  It biases toward tracking peers that store data with similar hashes to what you store in your own table, but it is intentionally sparse when keeping track of peers that cover hash-distant parts of the hash table. The distribution should be uniform, so you end up "ignoring" lots of those hash-distant peers. When deciding which of these peers to ignore, perhaps _physical_ distance could decide?

**Comment by mgrand** [comment]
Physical distance might be helpful, but nodes can provide false information
about their location. Network latency is more difficult to lie about unless
you have coordinated collusion for that purpose. Whether or not any of that
matters, I am not sure.
What I can tell you is that Kademlia has good performance over time
for chunks of data that are wanted in many places at once. It is what
bittorrent uses.

On Sat, Oct 5, 2024 at 8:12 PM Sean Cheatham ***@***.***>
wrote:

> 🥳 Yes! I'm basically hoping to achieve a DHT where the data is
> block/transaction information.
>
> I hadn't heard of Kademlia before but it looks super interesting! From my
> understanding, peers pick a random location within the hash table and store
> data with hashes "near" that location. Peers keep track of other peers and
> the sections of the hash table they are likely to store, but they bias
> toward tracking peers that are similar in the hash table to themselves.
> When requesting data that is close-by (from a hash perspective), it's very
> easy because you're so knowledgeable of nearby peers. When requesting data
> that's far away in the hash table, you find a peer in the rough ballpark
> and ask them for help because *they* know the area better than you. It
> might take multiple hops, but theoretically it should never take very many.
>
> Perhaps one way to tune Kademlia toward physical location is in the peer
> biasing. It biases toward tracking peers that store data with similar
> hashes to what you store in your own table, but it is intentionally sparse
> when keeping track of peers that cover hash-distant parts of the hash
> table. The distribution should be uniform, so you end up "ignoring" lots of
> those hash-distant peers. When deciding which of these peers to ignore,
> perhaps *physical* distance could decide?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/GiraffeChain/giraffe/discussions/38#discussioncomment-10855522>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEUGFXL3RKRHOTZAXSOIQ3Z2B56FAVCNFSM6AAAAABPIVYE6KVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTAOBVGU2TEMQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>


-- 
Mark Grand
(404)925 8265
Vision, Values, Verve:
* Visualize the end game with
* Values to lead ethically and with purpose; and
* Verve to execute with shameless audacity.
